Move mails

st3iny opened the pull request:
Adds an item to the Envelope and ThreadEnvelope context menus to move mails between mailboxes.

Feature was requested in #255.

dehnhardt commented:
Dear @st3iny,
I'm not the right one for a frontend review ;-) But what I noticed is that the changes in Envelope and ThreadEnvelope contain a lot of the equal code.
Wouldn't it make sense to create a component and include it in both places?
Holger

st3iny commented:
Thanks for your feedback. It makes sense to put the Modal into a separate component. I will push a new commit asap.

StCyr commented:
Yeah! Nice!

Thanks for this contribution. It's much awaited for me 👍 (I didn't realised there was a MailboxPicker component :-P)

@ChristophWurst, I've tested this branch on my server and it works great!

Here are my picky comments:

1. Shouldn't the "Move" action better be placed below the "Select" action?
2. What happens when you move a mail thread? Are all the thread's emails moved or only the one selected? Also, should we allow moving single emails from a thread? 
3. I would add this action to the "Multiselect" action bar too, won't you?

dehnhardt commented:
I tested this on my local system as well. I noticed that the moved message is not deleted in the source folder. I think it makes sense to do that in the frontend as well and not to wait for the synchronization. What do you think?

StCyr commented:
> I tested this on my local system as well. I noticed that the moved message is not deleted in the source folder. I think it makes sense to do that in the frontend as well and not to wait for the synchronization. What do you think?

I don't undertand your issue: for me messages look deleted directly (eg: enveloppes are removed from the enveloppe list and message is moved on the IMAP server)

ChristophWurst commented on src/store/actions.js:
```diff
@@ -707,4 +708,9 @@ export default {
 	},
+	async moveEnvelope({ dispatch }, { envelope, destMailboxId }) {
+		await moveMessage(envelope.databaseId, destMailboxId)
+		dispatch('syncEnvelopes', { mailboxId: envelope.mailboxId })
```
IMO the code that dispatches a `moveEnvelope` should subsequently trigger the sync, not the action itself

st3iny replied:
I agree. Ideally, only the affected envelopes should be updated but I couldn't get it to work reliably. Synching the whole mailbox maybe a bit overkill.

ChristophWurst commented:
> 2. What happens when you move a mail thread? Are all the thread's emails moved or only the one selected? Also, should we allow moving single emails from a thread?

Good point, though right now what you see in the list are always individual messages, not actual threads.

st3iny commented:
> 1. Shouldn't the "Move" action better be placed below the "Select" action?

I was referring to the original issue. jancborchardt wanted the entry to be placed below *Mark read/unread*. I guess we could place it above the delete entry to preserve consistency across all menus.

> 2. What happens when you move a mail thread? Are all the thread's emails moved or only the one selected? Also, should we allow moving single emails from a thread?

I forgot to test envelope threads. Currently, my code ignores threads and only moves single envelopes. When moving from the action in the mailbox list the whole thread should be moved. Moving single envelopes in the threading view should only move the selected envelope but not the whole thread.

> 3. I would add this action to the "Multiselect" action bar too, won't you?

Great idea.

> 4. The MailboxPicker component doesn't look very nice (but I think it's outside the scope of this PR, don't you think?)

The implementation/merge was expedited to be able to use it here. It should be improved in another PR.

st3iny commented:
I added an entry to the multiselect menu and moved the entry to be above the delete item in every menu. Feel free to review this PR.

ChristophWurst commented:
I've rebased and squashed the changes for you, hope that is OK :)

This works almost perfectly. Good job!

Three remarks

* The move modal should be more responsive in terms of height. If the window is small - e.g. on mobile - the *Move* button won't be visible
![Bildschirmfoto von 2020-09-29 10-06-14](https://user-images.githubusercontent.com/1374172/94531031-3a0a9f00-023c-11eb-89f7-d650e78743ac.png)
![Bildschirmfoto von 2020-09-29 10-06-19](https://user-images.githubusercontent.com/1374172/94531029-39720880-023c-11eb-97b3-4fd48f19700f.png)
* The actions menu should close when the move modal opens, otherwise it stays above the modal background and is still clickable.
* Occasionally the sync after the move make the browser VERY busy, so it even shows a warning and asks if the script should be terminated. I can't quite tell why that is. One time it was relatively fast, the other time it was terribly slow. This could be a nasty race condition. The time it was slow it fetched like the full mailbox as new messages, making the envelope list very long. Rendering that is expensive.

Thanks a lot :)

ChristophWurst commented:
From a release PoV it's too close to pack this into v1.5 but it could be one of the headline features of v1.6 :rocket:

dehnhardt commented:
> From a release PoV it's too close to pack this into v1.5 but it could be one of the headline features of v1.6

I really would love to see this in production as soon as possible. IMO not being able to move messages is a showstopper for professional use.

ChristophWurst commented:
I didn't say that we can't release this, right?

I just said it's too late for v1.5.

dehnhardt commented:
I didn't want to upset anyone - and by the way, I did not suspect that the fetaure would never be published. I just think that this is such an important feature that it would justify a small delay. Although I confess that I don't know if there is a fixed date for the release.

st3iny commented:
While trying to fix the responsiveness I decided that a dropdown is not really feasible inside a modal. Instead, I implemented a proper mailbox picker modal. I tried to match the appearance of the move/copy modal from the files app.

![MoveModal](https://user-images.githubusercontent.com/1479486/94838170-7f35f900-0415-11eb-8408-251f15afd1e1.png)

 Let me know what you think.

@dehnhardt Your MailboxPicker component works fine and has been a great starting point. Maybe we can use it somewhere else?

ChristophWurst commented:
> I didn't want to upset anyone - and by the way, I did not suspect that the fetaure would never be published. I just think that this is such an important feature that it would justify a small delay. Although I confess that I don't know if there is a fixed date for the release.

You didn't, don't worry. But we actually have plans for our releases. Tomorrow we'll push out v1.5 and for that it's too late to include a new *feature* now. There is no strict schedule after v1.5, hence a v1.6 can follow soon.

I just realized I forgot to set the data on the milestone, so you couldn't know that we run a fixed schedule for this release. My bad.

This is an important and awesome feature, yes, but we should take the time to make it a solid one.

ChristophWurst commented:
> Let me know what you think.

IMO it looks great!

Some minor remarks
* I think the icons should be a tad smaller
* The icons should possible have this 90% opacity only as we use in the sidebar. The 100% black is a bit much.
* Use the displayName and not the name of the mailbox (`INBOX` should be shown as `Inbox`)

st3iny commented:
I implemented your feedback. The icons are now smaller, slightly transparent and the mailbox names get translated correctly.

ChristophWurst commented:
> @dehnhardt Your MailboxPicker component works fine and has been a great starting point. Maybe we can use it somewhere else?

I'd say *this picker* should just replace the other one, no?

dehnhardt commented:
> > @dehnhardt Your MailboxPicker component works fine and has been a great starting point. Maybe we can use it somewhere else?
> 
> I'd say _this picker_ should just replace the other one, no?

I created the Mailboxpicker for the SieveFilter. For usability reasons I think I don't want a modal there - but I have to try. If it's ok for you I would like to keep it - or just maintain it in the SieveFilter branch again.

st3iny commented:
> Occasionally the sync after the move make the browser VERY busy, so it even shows a warning and asks if the script should be terminated. I can't quite tell why that is. One time it was relatively fast, the other time it was terribly slow. This could be a nasty race condition. The time it was slow it fetched like the full mailbox as new messages, making the envelope list very long. Rendering that is expensive.

I couldn't reproduce this issue because I only tested on a local imap server with a few mails on it. Maybe there is a race condition resulting from firing all the sync dispatches at the same time. A solution would be to wait for a sync to finish before starting the next one.

See the following code triggering the syncs:
```javascript
uniq(affectedMailboxes).forEach((mailboxId) => {
	this.$store.dispatch('syncEnvelopes', { mailboxId })
})
```

st3iny commented:
Ignore my last (deleted) comment.

I did some intensive testing and further improved the sync logic. Now it is very robust and fairly quick even on slow connections (tested via chromium devtools). I dived into the backend code and introduced some changes to `MailManager.php`.

The flow of moving a message currently is:
1. The user clicks move in the modal.
2. The php backend moves the imap message on the imap server.

If moving the imap message is successful:

3. The php backend also deletes the source message from the database cache.
4. The source envelope gets deleted from the vue store too.
5. After all envelopes have been moved successfully the sync is called once on the destination mailbox.

If there were any errors while moving multiple envelopes but at least one envelope was moved succesfully the sync of the destination mailbox is also triggered on closing the modal.

If you encounter a frozen move modal don't worry. This is a bug in the Modal component of [nextcloud-vue](https://github.com/nextcloud/nextcloud-vue). I fixed it locally and will file a PR there.

Let me know what you think. The move feature should be stable now.

st3iny commented:
Use [modal-freeze-patch.txt](https://github.com/nextcloud/mail/files/5328726/modal-freeze-patch.txt) for testing while the bug is being worked on.

jancborchardt commented:
Very nice! :) The home icon / breadcrumb does not make so much sense here, rather there should just be a back button to go a level back up.

st3iny commented:
The mail app and imap standard both allow deep nesting. Additionally, the user is being told about folders not mailboxes and that's why I tried to adhere to the move modal from the files app. In my opinion the home/crumbs thingy is helpful when navigating large mail accounts.

StCyr commented:
> The mail app and imap standard both allow deep nesting. Additionally, the user is being told about folders not mailboxes and that's why I tried to adhere to the move modal from the files app. In my opinion the home/crumbs thingy is helpful when navigating large mail accounts.

Agree: The breadcrumb might be useful and doesn't look annoying for the users

st3iny commented:
My fix over at nextcloud/vue got released in v2.6.9 and the dependency got bumped to v2.7.0 already. I rebased this branch once more and tested again. The move modal works fine now.

@ChristophWurst We can move on to reviewing/releasing this feature if there are no further issues.

ChristophWurst commented:
Unfortunately I still see the bad performance for the sync after the actual move with Firefox :/

![Bildschirmfoto von 2020-10-12 13-07-41](https://user-images.githubusercontent.com/1374172/95740013-114ac680-0c8c-11eb-8539-17572ecf0319.png)

ChristophWurst commented:
- [ ] UI/Accessibility bug: *Choose* button does not have a pointer cursor

ChristophWurst commented:
> Unfortunately I still see the bad performance for the sync after the actual move with Firefox :/

Did anyone except me ever experience this? Am I doing it wrong? :speak_no_evil:

StCyr commented:
> 
> 
> > Unfortunately I still see the bad performance for the sync after the actual move with Firefox :/
> 
> Did anyone except me ever experience this? Am I doing it wrong? 🙊

I tried once and didn't had any issue.

I'll try again today

ChristophWurst commented:
Btw @st3iny don't be confused about some weird file share you just received. I did this to create a guest account for you on our cloud so you can join our dev talk https://cloud.nextcloud.com/call/89474m7g :)

StCyr commented:
is there problems currently on cloud.nextcloud.com?


Erreur interne du serveur

Le serveur est incapable d'exécuter votre requête.

Si cela se reproduit, veuillez envoyer les détails techniques ci-dessous à l'administrateur du serveur.

Le fichier journal du serveur peut fournir plus de renseignements.
Renseignements techniques

    Adresse distante : 91.87.150.82
    ID de la demande : fsczGg8OR5CN2dXIxE8z

ChristophWurst commented:
> is there problems currently on cloud.nextcloud.com?

It was down in the morning shortly for a quick update. Should be ok now

StCyr commented:
> 
> 
> > is there problems currently on cloud.nextcloud.com?
> 
> It was down in the morning shortly for a quick update. Should be ok now

Still getting server error after login 😢

StCyr commented:
So, I tried moving some emails again here and didn't get any slowdowns here @ChristophWurst 

What I'm getting though is a 428 HTTP error upon a sync:

12:04:23.352
XHRPOSThttp://192.168.0.128/nextcloud/index.php/apps/mail/api/mailboxes/11/sync
[HTTP/1.1 428 Precondition required 119ms]

12:04:23.481
Could not sync envelopes: Request failed with status code 428 Error: Request failed with status code 428
    createError createError.js:16
    settle settle.js:17
    handleLoad xhr.js:69
    dispatchXhrRequest xhr.js:44
    xhrAdapter xhr.js:13
    dispatchRequest dispatchRequest.js:52
    promise callback*request Axios.js:61
    method Axios.js:86
    wrap bind.js:9
    _callee2$ MessageService.js:165
    Babel 8
    syncEnvelopes actions.js:581
    wrappedActionHandler vuex.esm.js:847
    dispatch vuex.esm.js:512
    boundDispatch vuex.esm.js:402
    onMove MoveModal.vue:145
    promise callback*onMove MoveModal.vue:144
    VueJS 33

Doesn't seem hurtful but better be fixed I guess

st3iny commented:
I rebased again last night and fixed some bugs related to moving multiple envelopes at once and moving thread envelopes. Moving multiple messages was completely broken and likely overloaded the whole frontend (somehow it did move ALL envelopes in the mailbox instead of just the selected).

@ChristophWurst Did Firefox freeze when you tried to select and move multiple enevelopes at once? This should be fixed now. Maybe you can test again?

----

> UI/Accessibility bug: Move button of the picker does not have a pointer cursor

The move button only has a pointer cursor when it is not disabled. Until a destination mailbox is selected it is disabled and shows the default cursor. The styles that are responsible for this are defined in server.css.

Can someone else confirm this? The styling is applied correctly on my local testing instance (version 20).

----

@StCyr Did you try the latest version? I pushed a commit last night that fixed some issues I found while testing with my day to day mail account. The line `MoveModal.vue:144` is just a curly brace in the current version :D

StCyr commented on src/components/MoveModal.vue:
```diff
@@ -0,0 +1,269 @@
+			})
+
+			Promise.all(promises).then(() => {
+				this.$store.dispatch('syncEnvelopes', { mailboxId: this.destMailboxId })
```
@st3iny : 

error "Could not sync envelopes: Request failed with status code 428 Error: Request failed with status code 428" most likely occurs here

StCyr replied:
'don't have the error anymore....

ChristophWurst reviewed (approved):
No more busy browser :heart_eyes:

StCyr commented:
Merged! 😄 🎈 

@ChristophWurst , I think following issues should be closed: https://github.com/nextcloud/mail/issues/1538, https://github.com/nextcloud/mail/issues/786, https://github.com/nextcloud/mail/issues/659, and https://github.com/nextcloud/mail/issues/254

Do you want me to do it?
